### ISSUE TITLE
problem with component references

### ISSUE THREAD

**Issue by dbacchet** [skypjack/entt#17]
Hi @skypjack, 

with EnTT 2.0.1, the reference to a component returned from calls like:
```C++
auto &a1   = ecs.assign<Alberto>(e1);
auto &a1_a = ecs.template get<Alberto>(e1); // this is just another alias for a1
```
is no more valid after I add the same component to other entities, like in the following example:
```C++
using ECS = entt::DefaultRegistry;

(...)

int main(int argc, char*argv[]) {
    ECS ecs;
    auto e1    = ecs.create();
    auto e2    = ecs.create();
    auto e3    = ecs.create();
    auto e4    = ecs.create();
    auto &a1   = ecs.assign<Alberto>(e1);
    auto &a1_a = ecs.template get<Alberto>(e1); // this is just another alias for a1
    // add an Alberto component to entity e2
    auto &t2   = ecs.assign<Alberto>(e2);

    // address of a1_b should be same of a1 and a1_a
    auto &a1_b = ecs.template get<Alberto>(e1); // this is different from a1 and a1_a
    std::cout << "Hi, I'm Alberto E1, and my address is: " << &a1 << std::endl;
    std::cout << "Hi, I'm Alberto E1, and my address is: " << &a1_a << std::endl;
    std::cout << "Hi, I'm Alberto E1, and my address is: " << &a1_b << std::endl;

    ecs.reset();
    return 0;
}
```
I would expect `a1`, `a1_a` and `a1_b` to be referencing the same object, instead with EnTT 2.0.1 `a1_b` points to a different memory address.

With EnTT v1, the same code works; the following is the output I get when I run the code with v1 and v2:
```
[~/dev/tmp/entt_testcase/v1] $ ./sample_entt_v1
Hi, I'm Alberto E1, and my address is: 0x1015c7000
Hi, I'm Alberto E1, and my address is: 0x1015c7000
Hi, I'm Alberto E1, and my address is: 0x1015c7000
[~/dev/tmp/entt_testcase/v1] $ cd ../v2
[~/dev/tmp/entt_testcase/v2] $ ./sample_entt_v2
Hi, I'm Alberto E1, and my address is: 0x7fe429c026e0
Hi, I'm Alberto E1, and my address is: 0x7fe429c026e0
Hi, I'm Alberto E1, and my address is: 0x7fe429c02860
[1]    90293 segmentation fault  ./sample_entt_v2
```

Attached there is a test case (I only tried with macOS 10.12) showing the behavior described above, and also reproducing the crash I described in #13.
[entt_testcase.zip](https://github.com/skypjack/entt/files/1405785/entt_testcase.zip)

As usual, thanks for your support and for sharing this library with the community!

**Comment by skypjack** [comment]
Hi @dbacchet 

>the reference to a component [...] is no more valid after I add the same component to other entities

Indeed, that's not exactly and the same was valid with `EnTT` v1. Let me explain what's changed between the two versions, why references aren't valid, what we can do to mitigate the problem (pros and cons) and why we can't solve it.

First of all, sparse sets and thus registry are containers. Use them how you would use containers from the standard library.  
As an example an `std::vector`. Would you keep around references returned from its `operator[]`? No, for they can be invalidated _by contract_. But you can access them if you know their positions in the array, right? An entity identifier is what we can call _the position of the components_ in the registry.

Components are stored within `std::vector`s in sparse sets. Each and every time an `std::vector` goes through a reallocation, references are invalidated.  
In `EnTT` v1 I gave vectors an initial size of 4096 and that's why with few entities you didn't notice that. However it didn't make much sense. I wasted a lot of space for a lot of types (think of a tag component, only one instance attached to an entity). I got rid of initial sizes in `EnTT` v2 and that's why the _issue_ popped out your side.

To mitigate it a bit I could allocate chunks of space in which to create components, so that reallocations of `std::vector`s won't affect component references.  
Unfortunately it has drawbacks:
* Components will be stored in semi-contiguous memory and iterations will incur in more cache misses
* Removing a component will still invalidate references (the last element of the array of components swaps with the one you are removing, so _it takes_ its address)

My two cents: it doesn't worth it.  
As a rule of thumb, you shouldn't store references around. Entity identifiers (returned from a `create` or by views) serve the same purpose and they aren't invalidated.  
Also views aren't invalidated if you add components during iterations for they don't work with references or raw pointers. Instead sparse sets provide them with custom iterators that abstract the underlying storage.  
In `EnTT` you should _play_ with entity identifiers and views, they are the _safe way_.

That being said, this is not a bug. It just works as intended and I cannot do much to help.

**Comment by dbacchet** [comment]
Hi @skypjack, that makes sense.
No need to change anything (well, maybe just make it explicit in the docs). On my side that only broke "artificial" examples (the unit tests) and had no impact in real code, where references are not saved at all. I'll just make sure that the unit tests use the right paradigm, it's an easy change :)

I'm closing this issue because is not a problem in EnTT
thanks!

**Comment by skypjack** [comment]
Good. Thank you.  
So, what about `EnTT` v2? I hope you enjoy the new features.  
Most of them are the results of your suggestions, you know? About the rest, I'm particularly proud of the standard vs persistent views and the design choice _pay for what you use_.  
What about? Requests for changes or features your side?

**Comment by LiliumAtratum** [comment]
@skypjack 
I just wanted to tell, that I faced exactly the same problem, albeit a bit more convoluted. I have a component with a std::function callback, with an arbitrary user code that can modify the registry. I ended up having the captured variables moved away while executing the said callback. It was really confusing at first...
However, since the reallocation is a necessity for the registry, I am _glad_ that you removed the initial 4096 size! Please do not bring it back :)
As it is now, I was able to track the problem while operating on merely 3 entities in a test environment. Now that I am aware of the problem, I can find a solution early in a development. If there was the initial allocation, I would probably hit the problem much, much later and it would be much harder to identify the problem and then deploy a fix.

**Comment by skypjack** [comment]
@LiliumAtratum what initial size of 4096? I don't think there was something like that before.
Btw did your function capture component references? If you need **truly** stable references for some reasons, you can have them with a custom pool.

**Comment by LiliumAtratum** [comment]
You said: _"In EnTT v1 I gave vectors an initial size of 4096"_ Maybe it's so old that you don't even remember :)
No, I didn't capture the component references, but the `std::function` itself is a component. During the execution of the said function it got moved away, and suddenly capture variables (by reference) weren't there.
I currently quick-solved my issue by having an additional redireciton, but I may check up on those custom pools, or maybe I'll find some other algorithmic solution to my problem. Thank you for the hint!
